src/ucp/proto/proto_am.inl

```diff
@@ -194,85 +175,125 @@ ucs_status_t ucp_do_am_zcopy_multi(uct_pending_req_t *self, uint8_t am_id_first,
-        if (status < 0) {
-            goto err; /* Failed */
-        }
 
-        UCS_PROFILE_REQUEST_EVENT(req, "am_zcopy_first", iov[0].length);
-        state->offset      += length_it;
-        ++req->send.uct_comp.count;
-        return UCS_INPROGRESS;
+        UCS_PROFILE_REQUEST_EVENT_CHECK_STATUS(req, "am_zcopy_first",
+                                               iov[0].length, status);
     } else if ((offset + max_middle < req->send.length) || flag_iov_mid) {
         /* Middle stage */
-        length_it = ucp_dt_iov_copy_uct(iov, &iovcnt, max_iov, state,
-                                        req->send.buffer, req->send.datatype, max_middle);
+        ucp_dt_iov_copy_uct(iov, &iovcnt, max_iov, &state, req->send.buffer,
+                            req->send.datatype, max_middle);
 
         status = uct_ep_am_zcopy(uct_ep, am_id_middle, (void*)hdr_middle,
                                  hdr_size_middle, iov, iovcnt, 0,
                                  &req->send.uct_comp);
-        if (status < 0) {
-            goto err; /* Failed */
-        }
 
-        UCS_PROFILE_REQUEST_EVENT(req, "am_zcopy_middle", iov[0].length);
-        state->offset      += length_it;
-        ++req->send.uct_comp.count;
-        return UCS_INPROGRESS;
+        UCS_PROFILE_REQUEST_EVENT_CHECK_STATUS(req, "am_zcopy_middle",
+                                               iov[0].length, status);
     } else {
         /* Last stage */
-        length_it = ucp_dt_iov_copy_uct(iov, &iovcnt, max_iov, state,
-                                        req->send.buffer, req->send.datatype,
-                                        req->send.length - offset);
+        ucp_dt_iov_copy_uct(iov, &iovcnt, max_iov, &state, req->send.buffer,
+                            req->send.datatype, req->send.length - offset);
 
         status = uct_ep_am_zcopy(uct_ep, am_id_last, (void*)hdr_middle,
                                  hdr_size_middle, iov, iovcnt, 0,
                                  &req->send.uct_comp);
-        if (status < 0) {
-            goto err; /* Failed */
-        }
-
-        UCS_PROFILE_REQUEST_EVENT(req, "am_zcopy_last", iov[0].length);
-        ucs_assert(req->send.length == (state->offset + length_it));
+        UCS_PROFILE_REQUEST_EVENT_CHECK_STATUS(req, "am_zcopy_last",
+                                               iov[0].length, status);
         if (status == UCS_OK) {
             complete(req, UCS_OK);
+            return UCS_OK;
+        }
+        ucp_request_send_state_advance(req, &state,
+                                       UCP_REQUEST_SEND_PROTO_ZCOPY_AM,
+                                       status);
+        return (status < 0) ? status : UCS_OK;
+    }
+
+    ucp_request_send_state_advance(req, &state,
+                                   UCP_REQUEST_SEND_PROTO_ZCOPY_AM,
+                                   status);
+
+    return (status < 0) ? status : UCS_INPROGRESS;
+}
+
+static UCS_F_ALWAYS_INLINE size_t
```

Comment:
i think this function should not be inline, zcopy is not fast path
can we somehow arrange that this function will be called only if short is not used? maybe add "pure" attribute?

Reply:
I measured performance of this PR, no effects on osu_latency and osu_mbw_mr. if we will add check for `short_max` outside this function it will look ugly. I think doing the calculations of all thresholds on initialization and save them in proto structures instead of header lengths would be better. the only one exception is iovs where we have dependency from input data type/len/count. It would be just usual function under `unlikely(is_iov)` branch.